### PR TITLE
Bump postgresql client to version 16

### PR DESCRIPTION
**Pull request by lars-fillmore**
Bumping postgresql client to a later version.  Official documentation notes that psql is backwards compatible.

_"If you want to use psql to connect to several servers of different major versions, it is recommended that you use the newest version of psql. Alternatively, you can keep around a copy of psql from each major version and be sure to use the version that matches the respective server. But in practice, this additional complication should not be necessary."_ 

https://www.postgresql.org/docs/current/app-psql.html

### CODE REVIEW

**Comment by pjonsson** [comment]
I've seen a PR on GDAL that updates its base image to Ubuntu 24.04 LTS. I don't know if that will end up/has ended up in what will be GDAL 3.9.0, but that would give a more recent Postgres client by just changing the version number like you've done.

**Comment by alexgleith** [comment]
I think you'll need to update the base image, from here: https://github.com/OSGeo/gdal/pkgs/container/gdal

That'll update Ubuntu underneath it.

**Comment by lars-fillmore** [comment]
GDAL 3,9 is still at release candidate 2.  I'll leave this as is rn and wait a bit for a full release.

File: index/Dockerfile [hunk 1]
```diff
@@ -14,21 +20,23 @@ RUN apt-get update \
...
     && apt-get autoclean \
     && apt-get autoremove \
     && rm -rf /var/lib/{apt,dpkg,cache,log}
 
+ENV VIRTUAL_ENV /virtualenv/python3.10
+ENV PATH /virtualenv/python3.10/bin:$PATH
+
 COPY requirements.txt constraints.txt version.txt /conf/
 
 RUN cat /conf/version.txt \
```
Comment (pjonsson) [inline index/Dockerfile:42]: Can't you just add ` --break-system-packages` after `install` in the next few lines and skip everything else?

Comment (lars-fillmore) [reply index/Dockerfile:42]: I'd prefer not to as this feature in Ubuntu 24 was introduced for a reason.  Other than a few extra lines of code, is there any other drawbacks from using `virtualenv`?

File: index/Dockerfile [hunk 2]
```diff
@@ -14,21 +20,23 @@ RUN apt-get update \
...
-        postgresql-client-14 \
+        postgresql-client-16 \
     # Cleanup
     && apt-get autoclean \
     && apt-get autoremove \
     && rm -rf /var/lib/{apt,dpkg,cache,log}
 
+ENV VIRTUAL_ENV /virtualenv/python3.12
+ENV PATH /virtualenv/python3.12/bin:$PATH
+
```
Comment (SpacemanPaul) [inline index/Dockerfile:39]: Do we need to set these manually? Isn't sourcing the activate script sufficient?  Or does the way Docker uses intermediate containers break that approach?

Comment (SpacemanPaul) [reply index/Dockerfile:39]: I guess if it's passing tests it's probably OK.

**Review by SpacemanPaul — APPROVED** [review]

File: index/Dockerfile (hunk 1 above)
Comment (lars-fillmore) [reply index/Dockerfile:42]: Thanks @SpacemanPaul Agree that the only real tangible benefit will be consistency on how the python runtime is used between both container and non-container environments, and hopefully make code maintenance easier in the future.

**Pull request merged**